Support trusted profile authentication in COS client

shahulsonhal opened the pull request:
**What this PR does / why we need it**:
Support trusted profile authentication in COS client

**Checklist**
- [x] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [x] Documentation added
- [x] Tests updated
- [ ] `CHANGELOG.md` updated
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/upgrading/_index.md`

shahulsonhal committed:
add documentation for COS storage (#8)

* Add documentation for COS as storage

* Address review comments

tareqmamari committed:
fix examples.md error

adityacs committed:
fix title in examples.md

tareqmamari committed:
fix: respect linting reports

This commit fixes few linting for both the code base related to COS storage
as well as the docs updates.

darxriggs committed:
chore: Fix typo in "tcp" (#8842)

chaudum committed:
Scheduler: Add query fairness control across multiple actors within a tenant (#8752)

**What this PR does / why we need it**:

This PR wires up the scheduler with the hierarchical queues. It is the last PR to implement https://github.com/grafana/loki/pull/8585.

When these changes are in place, the client performing query requests can control their QoS (query fairness) using the `X-Actor-Path` HTTP header. This header controls in which sub-queue of the tenant's scheduler queue the query request is enqueued. The place within the hierarchy where it is enqueued defines the probability with which the request gets dequeued.

A common use-case for this QoS control is giving each Grafana user within a tenant their fair share of query execution time.

Any documentation is still missing and will be provided by follow-up PRs.

**Special notes for your reviewer**:

```console
$ gotest -count=1 -v ./pkg/scheduler/queue/... -test.run=TestQueryFairness
=== RUN   TestQueryFairness
=== RUN   TestQueryFairness/use_hierarchical_queues_=_false
    dequeue_qos_test.go:109: duration actor a 2.007765568s
    dequeue_qos_test.go:109: duration actor b 2.209088331s
    dequeue_qos_test.go:112: total duration 2.209280772s
=== RUN   TestQueryFairness/use_hierarchical_queues_=_true
    dequeue_qos_test.go:109: duration actor b 605.283144ms
    dequeue_qos_test.go:109: duration actor a 2.270931324s
    dequeue_qos_test.go:112: total duration 2.271108551s
--- PASS: TestQueryFairness (4.48s)
    --- PASS: TestQueryFairness/use_hierarchical_queues_=_false (2.21s)
    --- PASS: TestQueryFairness/use_hierarchical_queues_=_true (2.27s)
PASS
ok  	github.com/grafana/loki/pkg/scheduler/queue	4.491s
```

```console
$ gotest -count=5 -v ./pkg/scheduler/queue/... -bench=Benchmark -test.run=^$ -benchtime=10000x -benchmem
goos: linux
goarch: amd64
pkg: github.com/grafana/loki/pkg/scheduler/queue
cpu: 11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz
BenchmarkGetNextRequest
BenchmarkGetNextRequest/without_sub-queues
BenchmarkGetNextRequest/without_sub-queues-8         	   10000	     29337 ns/op	    1600 B/op	     100 allocs/op
BenchmarkGetNextRequest/without_sub-queues-8         	   10000	     21348 ns/op	    1600 B/op	     100 allocs/op
BenchmarkGetNextRequest/without_sub-queues-8         	   10000	     21595 ns/op	    1600 B/op	     100 allocs/op
BenchmarkGetNextRequest/without_sub-queues-8         	   10000	     21189 ns/op	    1600 B/op	     100 allocs/op
BenchmarkGetNextRequest/without_sub-queues-8         	   10000	     21602 ns/op	    1600 B/op	     100 allocs/op
BenchmarkGetNextRequest/with_1_level_of_sub-queues
BenchmarkGetNextRequest/with_1_level_of_sub-queues-8 	   10000	     33770 ns/op	    2400 B/op	     200 allocs/op
BenchmarkGetNextRequest/with_1_level_of_sub-queues-8 	   10000	     33596 ns/op	    2400 B/op	     200 allocs/op
BenchmarkGetNextRequest/with_1_level_of_sub-queues-8 	   10000	     34432 ns/op	    2400 B/op	     200 allocs/op
BenchmarkGetNextRequest/with_1_level_of_sub-queues-8 	   10000	     33760 ns/op	    2400 B/op	     200 allocs/op
BenchmarkGetNextRequest/with_1_level_of_sub-queues-8 	   10000	     33664 ns/op	    2400 B/op	     200 allocs/op
BenchmarkGetNextRequest/with_2_levels_of_sub-queues
BenchmarkGetNextRequest/with_2_levels_of_sub-queues-8         	   10000	     71405 ns/op	    3200 B/op	     300 allocs/op
BenchmarkGetNextRequest/with_2_levels_of_sub-queues-8         	   10000	     59472 ns/op	    3200 B/op	     300 allocs/op
BenchmarkGetNextRequest/with_2_levels_of_sub-queues-8         	   10000	    117163 ns/op	    3200 B/op	     300 allocs/op
BenchmarkGetNextRequest/with_2_levels_of_sub-queues-8         	   10000	    106505 ns/op	    3200 B/op	     300 allocs/op
BenchmarkGetNextRequest/with_2_levels_of_sub-queues-8         	   10000	     64374 ns/op	    3200 B/op	     300 allocs/op
BenchmarkQueueRequest
BenchmarkQueueRequest-8                                       	   10000	    168391 ns/op	  320588 B/op	    1156 allocs/op
BenchmarkQueueRequest-8                                       	   10000	    166203 ns/op	  320587 B/op	    1156 allocs/op
BenchmarkQueueRequest-8                                       	   10000	    149518 ns/op	  320584 B/op	    1156 allocs/op
BenchmarkQueueRequest-8                                       	   10000	    219776 ns/op	  320583 B/op	    1156 allocs/op
BenchmarkQueueRequest-8                                       	   10000	    185198 ns/op	  320597 B/op	    1156 allocs/op
PASS
ok  	github.com/grafana/loki/pkg/scheduler/queue	64.648s
```

Signed-off-by: Christian Haudum <christian.haudum@gmail.com>

jeschkies committed:
Print duration in error messages with more readable. (#8816)

**What this PR does / why we need it**:
The old error messages would print only up to hours. E.g. `169h30s`.
This change will print it as `7d1h30s`. See
[model.Duration](https://github.com/prometheus/common/blob/66b493f42b757378fbad83e42b0ed441d425a17b/model/time.go#L259-L290)
for details.

**Checklist**
- [ ] Reviewed the
[`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md)
guide (**required**)
- [ ] Documentation added
- [x] Tests updated
- [ ] `CHANGELOG.md` updated
- [ ] Changes that require user attention or interaction to upgrade are
documented in `docs/sources/upgrading/_index.md`

dannykopping committed:
Ruler: rule evaluation jitter (#8848)

**What this PR does / why we need it**:
This PR introduces a configurable random sleep before each rule
evaluation to prevent contention on the query engine should multiple
runs execute concurrently. It's quite likely that rules will be
evaluated ~concurrently if they have the same interval.

ashwanthgoli committed:
LogQL: Fix panic when using `unpack` with empty line (#8853)

**What this PR does / why we need it**:

Fixes panic in unpack parser when processing empty lines

```
    /src/enterprise-logs/vendor/github.com/grafana/loki/pkg/logql/log/parser.go:583
github.com/grafana/loki/pkg/logql/log.(*UnpackParser).unpack(...)
panic: runtime error: index out of range [0] with length 0
```

Signed-off-by: Ashwanth Goli <iamashwanth@gmail.com>

chaudum committed:
Remove ./tools/fetch-tags invocation from Makefile (#8854)

**What this PR does / why we need it**:

This invocation causes problems when make targets are executed in the build image container (`BUILD_IN_CONTAINER=true`), e.g. the `yacc` target, because git tries to fetch tags from `origin` using ssh, causing
error message like these:

```
Host key verification failed.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
error: Could not fetch origin
```

This is because local ssh keys are not mounted into the build container.

In order not to break CI builds (https://github.com/grafana/loki/pull/8232) we need to manually fetch the tags as the first step in release pipelines. This is added to the Drone configuration.

Signed-off-by: Christian Haudum <christian.haudum@gmail.com>

JoaoBraveCoding committed:
ksonnet/loki: Add gRPC port to compactor mixin (#8855)

**What this PR does / why we need it**:

Problem: in https://github.com/grafana/loki/pull/7804 the possibility to use gRPC for communicating with the compactor for query time filtering of data requested for deletion was added, however, the Loki mixin still generates the compactor service without a gRPC port.

Solution: make the compactor service/deployment expose both an HTTP and gRPC port.

Signed-off-by: JoaoBraveCoding <jmarcal@redhat.com>

Suruthi-G-K committed:
Support trusted profile authentication in COS client

MasslessParticle committed:
Label selector optimizations (#8763)

This PR introduces the following label selector optimizations:
- any selector that is `=~<literal>` is converted to an
`Equals/NotEquals` matcher
- Use the greedy->non-greedy regex improvements that exist in line and
label filters
- Never run `.*` label matchers.

Prometheus regex matchers anchor the match to the beginning and end of
the line. That means matchers of the form `=~<literal_value>` are just
testing equality.

Benchmarks -- this is just a quick bench to demonstrate equality
matchers are faster than their equivalent regex matchers:
```
BenchmarkMatcherTypes/regex-8           27694834                53.74 ns/op            0 B/op          0 allocs/op
BenchmarkMatcherTypes/equals-8          627045850                2.153 ns/op           0 B/op          0 allocs/op
```

Red-GV committed:
operator: Update Dockerfile to go1.20 images (#8857)

btaani committed:
operator: Add support for rules configmap sharding (#7451)

dannykopping committed:
(chore): remove incorrect cache config preample (#8859)

With the addition of the [config generation
tool](https://github.com/grafana/loki/pull/7916), a comment prefix was
added for all cache types:

`Cache config for index entry writing.`
https://grafana.com/docs/loki/next/configuration/#cache_config

This appears in the description of each cache config option, and is
misleading.

dannykopping committed:
Ruler: remote rule evaluation hardening (#8785)

**What this PR does / why we need it**:
This PR is part 2 of 2 implementing remote rule evaluation.
See [part 1](https://github.com/grafana/loki/pull/8744) for more
context.

wtchangdm committed:
Loki: Add route_randomly to Redis options (#8852)

**What this PR does / why we need it**:

Current Redis client only reads from master node, which makes master
node under heavy pressure.

Usually, one or more read replicas are expected on production
environments.

This PR helps to increase the utilization of read replicas.

shahulsonhal committed:
Documentation for cos trusted profile auth

Add validation for trusted profile inputs
Add info logs to identify which auth is used by cos
Update test case for cos storage client
Update test case for trusted profile auth

Co-authored-by: tareqmamari <tariq.mamari@de.ibm.com>

periklis committed:
operator: Pin go version used in netlify env (#8861)

tareqmamari committed:
fix: make sure to register flags in storage factory

Without this registration, default config will be picked instead.

Signed-off-by: Shahul <shahulsonhal@gmail.com>

annelaurefroment committed:
update on Running Promtail on AWS EC2 tutorial (#8870)

Adding more information on what to do if `__meta_ec2_private_dns_name`
doesn't match the hostname.

MasslessParticle committed:
Short circuit parsers (#8724)

In many cases we only need a few labels from a line. Because most of our
parsers parse lines incrementally, we can stop parsing a line after we
have all the labels we want from it.

This pr uses `ParserHints` to keep track of the number of extracted
labels. It also provides a way for parsers to know when they should stop
parsing.

Notes:
- parsers had inconsistent ordering between the `ShouldExtract` call and
adding the `_extracted` label to duplicates. This PR makes appending
`_extracted` always happen before `ShouldExtract` to keep counts of what
is extracted compared to expected labels consistent.

Next Steps:
- When the user specifies a query with a grouping containing the
`_extracted` label but there is no duplicate between the passed line and
labels, short circuiting will not work. I'll address this in a follow up
PR.

Benchmarks:
To try and show a balanced view of what this buys us, this pr picks a
label our of the middle of a line. In a best case this might be much
better. In the worst case, we have to parse the whole line.

```
benchstat short_circuit_old.txt short_circuit_new.txt
name                               old time/op    new time/op    delta
KeyExtraction/json-8                  456ns ± 3%     256ns ± 2%  -43.84%  (p=0.000 n=9+10)
KeyExtraction/logfmt-8                347ns ± 4%     171ns ± 2%  -50.86%  (p=0.000 n=10+10)
KeyExtraction/logfmt-expression-8     552ns ± 2%     368ns ± 2%  -33.23%  (p=0.000 n=9+10)

name                               old alloc/op   new alloc/op   delta
KeyExtraction/json-8                  5.00B ± 0%     5.00B ± 0%     ~     (all equal)
KeyExtraction/logfmt-8                5.00B ± 0%     5.00B ± 0%     ~     (all equal)
KeyExtraction/logfmt-expression-8     16.0B ± 0%     16.0B ± 0%     ~     (all equal)

name                               old allocs/op  new allocs/op  delta
KeyExtraction/json-8                   1.00 ± 0%      1.00 ± 0%     ~     (all equal)
KeyExtraction/logfmt-8                 1.00 ± 0%      1.00 ± 0%     ~     (all equal)
KeyExtraction/logfmt-expression-8      2.00 ± 0%      2.00 ± 0%     ~     (all equal)
```

---------

Co-authored-by: J Stickler <julie.stickler@grafana.com>

JStickler committed:
Correcting typos merged in #8870. (#8873)

**Which issue(s) this PR fixes**:
Fixes typos introduced in #8870

tareqmamari commented on docs/sources/configuration/examples/13-COS-Trusted-Profile-Example.yaml:
```diff
@@ -0,0 +1,25 @@
+    region: <region>
+    auth_endpoint: <iam_endpoint_for_authentication>
+    cr_token_file_path: <path_to_compute_resource_token>
+    trusted_profile_name: <name_of_the_trusted_profile>
```
we need either the name or id, having both the name and id is confusing. Let's either use one of them, add a comment to explicitly mention that, or add another example for using the id instead of name.

tareqmamari commented on pkg/storage/chunk/client/ibmcloud/cos_object_client_test.go:
```diff
@@ -518,3 +529,74 @@ func Test_APIKeyAuth(t *testing.T) {
+	tokenType := "Bearer"
+	resp := "testGet"
+
+	cosSvr := cosServer(testToken, tokenType, resp)
```
could we rename the srv to clearly mention that this is nothing but a fake server?

adityacs commented on pkg/storage/chunk/client/ibmcloud/cos_object_client_test.go:
```diff
@@ -518,3 +529,74 @@ func Test_APIKeyAuth(t *testing.T) {
+		}
+
+		w.WriteHeader(http.StatusOK)
+		w.Write(data)
```
Better to handle the `Write` error and say `StatusUnauthorized` if any error.

shahulsonhal replied:
It's something we can't do, the problem is if we call `WriteHeader` after the `Write` they will show a warning. As far as I understood it's okay to skip error check in this case. What maximum we can do is to just log the error, since it's test casae I don't think we have to do that.

adityacs replied:
My doubt is that we might get lint error as we are not capturing the error and handling it. Try running lint once from Loki Makefile.

tareqmamari replied:
Agree with @adityacs , that will make the linting fail.

shahulsonhal replied:
Linter was not complaining for me when I ran it locally. Also I have seen that people adding exception for function like `w.Write(data)` and  `fmt.Fprintln(w, resp)` since sometimes checking error for them deos not make any sense.

As I explained above if we call `WriteHeader` after the `Write`, we will get a warining when we run the test case which says that `WriteHeader` statement is unnecessary. Eg.., below:

```
2023/03/23 13:42:50 http: superfluous response.WriteHeader call from github.com/grafana/loki/pkg/storage/chunk/client/ibmcloud.authServer.func1 (cos_object_client_test.go:600
```

tareqmamari replied:
if you open the PR to Loki, it would complain, the same issue happened with the other PR.

shahulsonhal replied:
We are checking the error and logging it now.

tareqmamari commented on pkg/storage/chunk/client/ibmcloud/trusted_profile_authentication_provider.go:
```diff
@@ -0,0 +1,159 @@
+	}, nil
+}
+
+// IsExpired ...
```
cleanup the comment

tareqmamari commented on pkg/storage/chunk/client/ibmcloud/trusted_profile_authentication_provider.go:
```diff
@@ -0,0 +1,159 @@
+		return credentials.Value{ProviderName: p.providerName}, p.ErrorStatus
+	}
+
+	tokenValue, err := p.authenticator.GetToken()
```
add a comment to explain that this gets a token or generate a new one if expired or there is none that are already retrieved.

adityacs replied:
+1

tareqmamari commented on pkg/storage/chunk/client/ibmcloud/trusted_profile_authentication_provider.go:
```diff
@@ -0,0 +1,159 @@
+		level.Debug(log.Logger).Log("msg", provider.ErrorStatus)
+
+		return provider
+	} else {
```
no need for the else here.

Suruthi-G-K replied:
Tareq, if the CR token filepath is given, in the else part we are checking if the CR token file exist or not by calling the os.Stat method. 
If the file does not exist we are returning provider

tareqmamari replied:
@Suruthi-G-K since we do a return in line 67, there is no need for the else statement here.

Suruthi-G-K replied:
@tareqmamari i misunderstood. Got it now.

Red-GV committed:
operator: Remove mutations to non-updatable statefulset fields (#8875)

chaudum committed:
Rename LeafQueue to TreeQueue (#8856)

**What this PR does / why we need it**:

`TreeQueue` is the semantically more correct term for this type of queue.

Signed-off-by: Christian Haudum <christian.haudum@gmail.com>

tareqmamari commented on pkg/storage/chunk/client/ibmcloud/trusted_profile_authentication_provider.go:
```diff
@@ -0,0 +1,159 @@
+	tokenType                  = "Bearer"
+)
+
+// TrustedProfileProvider Struct
```
this needs further simplification:
`TrustedProfileProvider implements Provider interface from https://github.com/IBM/ibm-cos-sdk-go`

tareqmamari commented on pkg/storage/chunk/client/ibmcloud/trusted_profile_authentication_provider.go:
```diff
@@ -0,0 +1,159 @@
+	ErrorStatus error
+}
+
+// NewTrustedProfileProvider allows the creation of a custom IBM IAM Provider
```
this one as well

jeschkies committed:
Document how to install Loki and Promtail using APT/DNF. (#8841)

**What this PR does / why we need it**:
With #6456 we started providing native packages of Loki, Promtail and
logcli. They've been added to Grafana's repository in the meantime.

In order to spread the knowledge of the packages and the repo it should
be documented.

**Which issue(s) this PR fixes**:
Fixes #52


**Checklist**
- [ ] Reviewed the
[`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md)
guide (**required**)
- [x] Documentation added
- [ ] Tests updated
- [ ] `CHANGELOG.md` updated
- [ ] Changes that require user attention or interaction to upgrade are
documented in `docs/sources/upgrading/_index.md`

---------

Co-authored-by: René Scheibe <rene.scheibe@gmail.com>
Co-authored-by: J Stickler <julie.stickler@grafana.com>

periklis committed:
operator: Refactor all type validations into own package (#8878)

chaudum committed:
Extract scheduler queue metrics into separate field

This allows for easier passing of the metrics to the scheduler
instantiation.

Signed-off-by: Christian Haudum <christian.haudum@gmail.com>

periklis committed:
operator: Break the API types out into their own module (#8863)

tareqmamari commented on pkg/storage/chunk/client/ibmcloud/cos_object_client.go:
```diff
@@ -163,13 +185,36 @@ func validate(cfg COSConfig) error {
 }
 
 func getCreds(cfg COSConfig) *credentials.Credentials {
+	if cfg.CRTokenFilePath != "" {
```
what's about we convert the following if conditions to a switch?

adityacs replied:
+1

tareqmamari commented on pkg/storage/chunk/client/ibmcloud/cos_object_client_test.go:
```diff
@@ -194,7 +195,7 @@ func Test_COSConfig(t *testing.T) {
 		},
 		{
-			"Access key ID and Secret Access key and APIKey is empty",
+			"Access key ID and Secret Access key and APIKey and CR token file path is empty",
```
we need to correct this sentence.
Access key ID, Secret Access key, APIKey and CR token file path are empty.

tareqmamari commented on pkg/storage/chunk/client/ibmcloud/cos_object_client_test.go:
```diff
@@ -219,7 +220,18 @@ func Test_COSConfig(t *testing.T) {
 		},
 		{
-			"valid config",
+			"CR token file path with empty trusted profile name and trusted profile ID",
```
CR token file path with empty trusted profile name and ID.

tareqmamari commented on pkg/storage/chunk/client/ibmcloud/cos_object_client_test.go:
```diff
@@ -241,13 +253,35 @@ func Test_COSConfig(t *testing.T) {
+				Region:             "dummy",
+				AuthEndpoint:       "dummy",
+				CRTokenFilePath:    "dummy",
+				TrustedProfileName: "test",
```
let's use either the name or id

tareqmamari commented on pkg/storage/chunk/client/ibmcloud/trusted_profile_authentication_provider.go:
```diff
@@ -0,0 +1,159 @@
+	} else {
+		if _, err := os.Stat(crTokenFilePath); errors.Is(err, os.ErrNotExist) {
+			provider.ErrorStatus = awserr.New("crTokenFileNotFound", "no such file", err)
+			level.Debug(log.Logger).Log("msg", "no such file", "err", err)
```
which file? if you get the error in the logs, will we be able to identify which file we are talking about?

shahulsonhal replied:
Yes, it will look like as below in logs:
```
crTokenFileNotFound: no such file
caused by: stat /Users/shahul/go/src/loki/tokens: no such file or directory
error creating index client
github.com/grafana/loki/pkg/storage.(*store).storeForPeriod
.
.
.
{"caller":"trusted_profile_authentication_provider.go:72","err":"stat /Users/shahul/go/src/loki/tokens: no such file or directory","level":"debug","msg":"no such file","ts":"2023-03-23T16:56:36.39803Z"}
```

tareqmamari reviewed (changes requested):
I left few comments

JoaoBraveCoding committed:
mixins: Normalize headless service name for query-frontend/scheduler (#8880)

**What this PR does / why we need it**:
Mixins in general are rather complex pieces of code to consume. When
trying to use the loki mixins it wasn't apparent at first that I could
use a headless service for DNS SRV discovery for both query_frontend and
query_scheduler, especially because for frontend the service is named
`query_frontend_headless_service` and for the scheduler, it's named
`query_scheduler_service_discovery`. Furthermore, the query_frontend
mixin provides a non-headless version of the service while
query_scheduler doesn't. This PR aims at normalizing this for end users,
both headless services are now named `query_frontend_headless_service`
and `query_scheduler_headless_service` and both also provide a
non-headless version of the service.

sandeepsukhani committed:
tsdb: sample chunk info from tsdb index to limit the amount of chunkrefs we read from index (#8742)

**What this PR does / why we need it**:
Previously we used to read the info of all the chunks from the index and
then filter it out in a layer above within the tsdb code. This wastes a
lot of resources when there are too many chunks in the index, but we
just need a few of them based on the query range.

Before jumping into how and why I went with chunk sampling, here are
some points to consider:
* Chunks in the index are sorted by the start time of the chunk. Since
this does not tell us much about the end time of the chunks, we can only
skip chunks that start after the end time of the query, which still
would make us process lots of chunks when the query touches chunks that
are near the end of the table boundary.
* Data is written to tsdb with variable length encoding. This means we
can't skip/jump chunks since each chunk info might vary in the number of
bytes we write.

Here is how I have implemented the sampling approach:
* Chunks are sampled considering their end times from the index and
stored in memory.
* Here is how `chunkSample` is defined:
```
type chunkSample struct {
	largestMaxt   int64 // holds largest chunk end time we have seen so far. In other words all the earlier chunks have maxt <= largestMaxt
	idx           int   // index of the chunk in the list which helps with determining position of sampled chunk
	offset        int   // offset is relative to beginning chunk info block i.e after series labels info and chunk count etc
	prevChunkMaxt int64 // chunk times are stored as deltas. This is used for calculating mint of sampled chunk
}
```
* When a query comes in, we will find `chunkSample`, which has the
largest "largestMaxt" that is less than the given query start time. In
other words, find a chunk sample which skips all/most of the chunks that
end before the query start time.
* Once we have found a chunk sample which skips all/most of the chunks
that end before the query start, we will sequentially go through chunks
and consider only the once that overlap with the query range. We will
stop processing chunks as soon as we see a chunk that starts after the
end time of the query since the chunks are sorted by start time.
* Sampling of chunks is done lazily for only the series that are
queried, so we do not waste any resources on sampling series that are
not queried.
* To avoid sampling too many chunks, I am sampling chunks at `1h` steps
i.e given a sampled chunk with chunk end time `t`, the next chunk would
be sampled with end time >= `t + 1h`. This means typically, we should
have ~28 chunks sampled for each series queried from each index file,
considering 2h default chunk length and chunks overlapping multiple
tables.

Here are the benchmark results showing the difference it makes:
```
benchmark                              old ns/op     new ns/op     delta
BenchmarkTSDBIndex_GetChunkRefs-10     12420741      4764309       -61.64%
BenchmarkTSDBIndex_GetChunkRefs-10     12412014      4794156       -61.37%
BenchmarkTSDBIndex_GetChunkRefs-10     12382716      4748571       -61.65%
BenchmarkTSDBIndex_GetChunkRefs-10     12391397      4691054       -62.14%
BenchmarkTSDBIndex_GetChunkRefs-10     12272200      5023567       -59.07%

benchmark                              old allocs     new allocs     delta
BenchmarkTSDBIndex_GetChunkRefs-10     345653         40             -99.99%
BenchmarkTSDBIndex_GetChunkRefs-10     345653         40             -99.99%
BenchmarkTSDBIndex_GetChunkRefs-10     345653         40             -99.99%
BenchmarkTSDBIndex_GetChunkRefs-10     345653         40             -99.99%
BenchmarkTSDBIndex_GetChunkRefs-10     345653         40             -99.99%

benchmark                              old bytes     new bytes     delta
BenchmarkTSDBIndex_GetChunkRefs-10     27286536      6398855       -76.55%
BenchmarkTSDBIndex_GetChunkRefs-10     27286571      6399276       -76.55%
BenchmarkTSDBIndex_GetChunkRefs-10     27286566      6400699       -76.54%
BenchmarkTSDBIndex_GetChunkRefs-10     27286561      6399158       -76.55%
BenchmarkTSDBIndex_GetChunkRefs-10     27286580      6399643       -76.55%
```

**Checklist**
- [x] Tests updated

jeschkies committed:
Define `RequiredLabels` query limit. (#8851)

**What this PR does / why we need it**:
Some end-users can impose great workload on a cluster by selecting too
many streams in their queries. We should be able to limit them.

Therefore we introduce a new limit `RequiredLabelMatchers` which list
label names that must be included in the stream selectors.

The implementation follows the same approach as for max query limit.

**Which issue(s) this PR fixes**:
Fixes #8745

**Checklist**
- [ ] Reviewed the
[`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md)
guide (**required**)
- [x] Documentation added
- [x] Tests updated
- [x] `CHANGELOG.md` updated
- [ ] Changes that require user attention or interaction to upgrade are
documented in `docs/sources/upgrading/_index.md`

salvacorts committed:
Max bytes read limit (#8670)

**What this PR does / why we need it**:

This PR implements two new per-tenant limits that are enforced on log
and metric queries (both range and instant) when TSDB is used:

- `max_query_bytes_read`: Refuse queries that would read more than the
configured bytes here. Overall limit regardless of splitting/sharding.
The goal is to refuse queries that would take too long. The default
value of 0 disables this limit.

- `max_querier_bytes_read`: Refuse queries in which any of their
subqueries after splitting and sharding would read more than the
configured bytes here. The goal is to avoid a querier from running a
query that would load too much data in memory and can potentially get
OOMed. The default value of 0 disables this limit.

These new limits can be configured per tenant and per query (see
https://github.com/grafana/loki/pull/8727).

The bytes a query would read are estimated through TSDB's index stats.
Even though they are not exact, they are good enough to have a rough
estimation of whether a query is too big to run or not. For more details
on this refer to this discussion in the PR:
https://github.com/grafana/loki/pull/8670#discussion_r1124858508.

Both limits are implemented in the frontend. Even though we considered
implementing `max_querier_bytes_read` in the querier, this way, the
limits for pre and post splitting/sharding queries are enforced close to
each other on the same component. Moreover, this way we can reduce the
number of index stats requests issued to the index gateways by reusing
the stats gathered while sharding the query.

With regard to how index stats requests are issued:
- We parallelize index stats requests by splitting them into queries
that span up to 24h since our indices are sharded by 24h periods. On top
of that, this prevents a single index gateway from processing a single
huge request like `{app=~".+"} for 30d`.
- If sharding is enabled and the query is shardable, for
`max_querier_bytes_read`, we re-use the stats requests issued by the
sharding ware. Specifically, we look at the [bytesPerShard][1] to
enforce this limit.

Note that once we merge this PR and enable these limits, the load of
index stats requests will increase substantially and we may discover
bottlenecks in our index gateways and TSDB. After speaking with @owen-d,
we think it should be fine as, if needed, we can scale up our index
gateways and support caching index stats requests.

Here's a demo of this working:
<img width="1647" alt="image"
src="https://user-images.githubusercontent.com/8354290/226918478-d4b6c2fd-de4d-478a-9c8b-e38fe148fa95.png">

<img width="1647" alt="image"
src="https://user-images.githubusercontent.com/8354290/226918798-a71b1db8-ea68-4d00-933b-e5eb1524d240.png">


**Which issue(s) this PR fixes**:
This PR addresses https://github.com/grafana/loki-private/issues/674.

**Special notes for your reviewer**:

- @jeschkies has reviewed the changes related to query-time limits.
- I've done some refactoring in this PR:
- Extracted logic to get stats for a set of matches into a new function
[getStatsForMatchers][2].
- Extracted the _Handler_ interface implementation for
[queryrangebase.roundTripper][3] into a new type
[queryrangebase.roundTripperHandler][4]. This is used to create the
handler that skips the rest of configured middlewares when sending an
index stat quests ([example][5]).

**Checklist**
- [x] Reviewed the
[`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md)
guide (**required**)
- [x] Documentation added
- [x] Tests updated
- [x] `CHANGELOG.md` updated
- [ ] Changes that require user attention or interaction to upgrade are
documented in `docs/sources/upgrading/_index.md`


[1]:
https://github.com/grafana/loki/blob/ff847305afaf7de5eb56436f3683773e88701075/pkg/querier/queryrange/shard_resolver.go#L179-L186

[2]:
https://github.com/grafana/loki/blob/ff847305afaf7de5eb56436f3683773e88701075/pkg/querier/queryrange/shard_resolver.go#L72

[3]:
https://github.com/grafana/loki/blob/3d2fff3a2d416a48a73346a53ba7499b0eeb67f7/pkg/querier/queryrange/queryrangebase/roundtrip.go#L124

[4]:
https://github.com/grafana/loki/blob/3d2fff3a2d416a48a73346a53ba7499b0eeb67f7/pkg/querier/queryrange/queryrangebase/roundtrip.go#L163

[5]:
https://github.com/grafana/loki/blob/f422e0a52b743a11209b8276510feb2ab8241486/pkg/querier/queryrange/roundtrip.go#L521

DylanGuedes committed:
Loki: Add more spans to write path (#8888)

**What this PR does / why we need it**:
Add new spans to our write path to better determine which operations are
holding our write performance.

bilbof committed:
Canary: support filtering / parsing logs with LogQL (#8871)

**What this PR does / why we need it**:

The canary logs can be shipped to Loki in various formats. This allows
users of the canary to parse out the canaries logs from any format e.g.
nested json.

The alternative is to do something different for the canary logs to
prevent the logs being wrapped / manipulated (e.g. write to Loki
directly, or use a different log pipeline) which limits the benefits of
the canary.

I've confirmed this works locally.

**Which issue(s) this PR fixes**:
Fixes #7775

**Checklist**
- [x] Reviewed the
[`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md)
guide (**required**)
- [ ] Documentation added
- [ ] Tests updated
- [ ] `CHANGELOG.md` updated
- [ ] Changes that require user attention or interaction to upgrade are
documented in `docs/sources/upgrading/_index.md`

bboreham committed:
WAL: remove ePool that is unused (#8669)

**What this PR does / why we need it**:

Nothing is ever fetched from the pool - `GetEntries()` is never called.
And the loop to call `PutEntries()` follows a call where the slice it
ranges over is reset to empty, so that loop never added anything to the
pool.

**Checklist**
- [x] Reviewed the
[`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md)
guide (**required**)
- NA Documentation added
- NA Tests updated
- NA `CHANGELOG.md` updated
- NA Changes that require user attention or interaction to upgrade are
documented in `docs/sources/upgrading/_index.md`

bboreham committed:
Iterators: re-implement mergeEntryIterator using loser.Tree for performance (#8637)

**What this PR does / why we need it**:

Building on #8351, this re-implements `mergeEntryIterator` using
`loser.Tree`; the benchmark says it goes much faster but uses a bit more
memory (while building the tree).

```
name                       old time/op    new time/op    delta
SortIterator/merge_sort-4    10.7ms ± 4%     2.9ms ± 2%  -72.74%  (p=0.008 n=5+5)

name                       old alloc/op   new alloc/op   delta
SortIterator/merge_sort-4    11.2kB ± 0%    21.7kB ± 0%  +93.45%  (p=0.008 n=5+5)

name                       old allocs/op  new allocs/op  delta
SortIterator/merge_sort-4      6.00 ± 0%      7.00 ± 0%  +16.67%  (p=0.008 n=5+5)
```

The implementation is very different: rather than relying on iterators
supporting `Peek()`, `mergeEntryIterator` now pulls items into its
buffer until it finds one with a different timestamp or stream, and
always works off what is in the buffer.
The comment `"[we] pop the ones whose common value occurs most often."`
did not appear to match the previous implementation, and no attempt was
made to match this comment.

A `Push()` function was added to `loser.Tree` to support live-streaming.
This works by finding or making an empty slot, then re-running the
initialize function to find the new winner. A consequence is that the
previous "winner" value is lost after calling
`Push()`, and users must call `Next()` to see the next item.

A couple of tests had to be amended to avoid assuming particular
behaviour of the implementation; I recommend that reviewers consider
these closely.

**Checklist**
- [x] Reviewed the
[`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md)
guide (**required**)
- NA Documentation added
- [x] Tests updated
- NA `CHANGELOG.md` updated
- NA Changes that require user attention or interaction to upgrade are
documented in `docs/sources/upgrading/_index.md`

andriikushch committed:
Promtail: Add a new target for the Azure Event Hub (#8787)

**What this PR does / why we need it**:

We want to allow receiving logs from the Azure Cloud. To solve this
problem, we need to be able to configure Promtail to consume logs from
the Azure Events Hub.

To achieve this goal, we need to add a new configuration option to the
scrape configuration in Promtail and implement the component that, given
this configuration, can connect to the Azure Event Hub, parse received
data, and prepare it to be stored in Loki.


**Which issue(s) this PR fixes**:
Fix [#8788](https://github.com/grafana/loki/issues/8788)

# **Special notes for your reviewer**:

With the current implementation, I would like to leverage Azure Events
Hub's compatibility with a Kafka protocol and reuse code for a Kafka
target.

### Why I decided to create a new target instead of just reusing the
Kafka target:
1. Configuration for Kafka in this scenario requires fewer parameters.
For example, version and auth should have some specific values.
2. Message handling is different for Azure Logs. In addition to what we
are doing for the Kafka target, we want to fix JSON and split incoming
messages into multiple log lines before propagating it further.

### From the implementation perspective, there are two goals:
- Extract interfaces that are not Kafka configuration specific from the
Kafka target module. Splitting Kafka config parsing and target creation.

- To make the message handler/parser injectable or configurable: we
would like to fix JSON and split a message into multiple messages.


**Checklist**
- [x] Reviewed the
[`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md)
guide (**required**)
- [x] Documentation added
- [x] Tests updated
- [x] `CHANGELOG.md` updated
- [ ] Changes that require user attention or interaction to upgrade are
documented in `docs/sources/upgrading/_index.md`

---------

Co-authored-by: Pablo <2617411+thepalbi@users.noreply.github.com>
Co-authored-by: Paschalis Tsilias <tpaschalis@users.noreply.github.com>
Co-authored-by: J Stickler <julie.stickler@grafana.com>

dependabot[bot] committed:
Bump actions/setup-go from 3 to 4 (#8837)

Bumps [actions/setup-go](https://github.com/actions/setup-go) from 3 to
4.
<details>
<summary>Release notes</summary>
<p><em>Sourced from <a
href="https://github.com/actions/setup-go/releases">actions/setup-go's
releases</a>.</em></p>
<blockquote>
<h2>v4.0.0</h2>
<p>In scope of release we enable cache by default. The action won’t
throw an error if the cache can’t be restored or saved. The action will
throw a warning message but it won’t stop a build process. The cache can
be disabled by specifying <code>cache: false</code>.</p>
<pre lang="yaml"><code>steps:
  - uses: actions/checkout@v3
  - uses: actions/setup-go@v4
    with:
      go-version: ‘1.19’
  - run: go run hello.go
</code></pre>
<p>Besides, we introduce such changes as</p>
<ul>
<li><a
href="https://redirect.github.com/actions/setup-go/pull/305">Allow to
use only GOCACHE for cache</a></li>
<li><a href="https://redirect.github.com/actions/setup-go/pull/315">Bump
json5 from 2.2.1 to 2.2.3</a></li>
<li><a href="https://redirect.github.com/actions/setup-go/pull/323">Use
proper version for primary key in cache</a></li>
<li><a
href="https://redirect.github.com/actions/setup-go/pull/351">Always add
Go bin to the PATH</a></li>
<li><a href="https://redirect.github.com/actions/setup-go/pull/350">Add
step warning if go-version input is empty</a></li>
</ul>
<h2>Add support for stable and oldstable aliases</h2>
<p>In scope of this release we introduce aliases for the
<code>go-version</code> input. The <code>stable</code> alias instals the
latest stable version of Go. The <code>oldstable</code> alias installs
previous latest minor release (the stable is 1.19.x -&gt; the oldstable
is 1.18.x).</p>
<h3>Stable</h3>
<pre lang="yaml"><code>steps:
  - uses: actions/checkout@v3
  - uses: actions/setup-go@v3
    with:
      go-version: 'stable'
  - run: go run hello.go
</code></pre>
<h3>OldStable</h3>
<pre lang="yaml"><code>steps:
  - uses: actions/checkout@v3
  - uses: actions/setup-go@v3
    with:
      go-version: 'oldstable'
  - run: go run hello.go
</code></pre>
<h2>Add support for go.work and pass the token input through on
GHES</h2>
<p>In scope of this release we added <a
href="https://redirect.github.com/actions/setup-go/pull/283">support for
go.work file to pass it in go-version-file input</a>.</p>
<pre lang="yaml"><code>steps:
  - uses: actions/checkout@v3
  - uses: actions/setup-go@v3
&lt;/tr&gt;&lt;/table&gt; 
</code></pre>
</blockquote>
<p>... (truncated)</p>
</details>
<details>
<summary>Commits</summary>
<ul>
<li><a
href="https://github.com/actions/setup-go/commit/4d34df0c2316fe8122ab82dc22947d607c0c91f9"><code>4d34df0</code></a>
Update configuration files (<a
href="https://redirect.github.com/actions/setup-go/issues/348">#348</a>)</li>
<li><a
href="https://github.com/actions/setup-go/commit/fdc0d672a12b8fc05ef4c549c3fd5b0f03ebeb4c"><code>fdc0d67</code></a>
Add Go bin if go-version input is empty (<a
href="https://redirect.github.com/actions/setup-go/issues/351">#351</a>)</li>
<li><a
href="https://github.com/actions/setup-go/commit/ebfdf6ac95bce0e7e743a60aba7adf59b400885b"><code>ebfdf6a</code></a>
add warning if go-version is empty (<a
href="https://redirect.github.com/actions/setup-go/issues/350">#350</a>)</li>
<li><a
href="https://github.com/actions/setup-go/commit/b27d76912e7cb72b7262b5be2ec903b6cfaf358a"><code>b27d769</code></a>
fix lockfileVersion (<a
href="https://redirect.github.com/actions/setup-go/issues/349">#349</a>)</li>
<li><a
href="https://github.com/actions/setup-go/commit/c51a7207680f1bee06f4a751214aab70667f9e25"><code>c51a720</code></a>
Enable caching by default with default input (<a
href="https://redirect.github.com/actions/setup-go/issues/332">#332</a>)</li>
<li><a
href="https://github.com/actions/setup-go/commit/6b848af622919e18822928df24a57044dc6ea9ab"><code>6b848af</code></a>
Merge pull request <a
href="https://redirect.github.com/actions/setup-go/issues/343">#343</a>
from akv-platform/reusable-workflow</li>
<li><a
href="https://github.com/actions/setup-go/commit/12741cc20952a981105f1dc8366d69f1d24ed472"><code>12741cc</code></a>
Format update-config-files.yml</li>
<li><a
href="https://github.com/actions/setup-go/commit/7a77a6aab66a15178e313f5908f5ed9b8500894c"><code>7a77a6a</code></a>
Merge branch 'main' into reusable-workflow</li>
<li><a
href="https://github.com/actions/setup-go/commit/42a0cc8e1491c14a0e9486c3a895d2124f51af1a"><code>42a0cc8</code></a>
Add update-config-files.yml</li>
<li><a
href="https://github.com/actions/setup-go/commit/7406d654ad8f517886f6c77135494c2b99355381"><code>7406d65</code></a>
Add and configure ESLint and update configuration for Prettier (<a
href="https://redirect.github.com/actions/setup-go/issues/341">#341</a>)</li>
<li>Additional commits viewable in <a
href="https://github.com/actions/setup-go/compare/v3...v4">compare
view</a></li>
</ul>
</details>
<br />


[![Dependabot compatibility
score](https://dependabot-badges.githubapp.com/badges/compatibility_score?dependency-name=actions/setup-go&package-manager=github_actions&previous-version=3&new-version=4)](https://docs.github.com/en/github/managing-security-vulnerabilities/about-dependabot-security-updates#about-compatibility-scores)

Dependabot will resolve any conflicts with this PR as long as you don't
alter it yourself. You can also trigger a rebase manually by commenting
`@dependabot rebase`.

[//]: # (dependabot-automerge-start)
[//]: # (dependabot-automerge-end)

---

<details>
<summary>Dependabot commands and options</summary>
<br />

You can trigger Dependabot actions by commenting on this PR:
- `@dependabot rebase` will rebase this PR
- `@dependabot recreate` will recreate this PR, overwriting any edits
that have been made to it
- `@dependabot merge` will merge this PR after your CI passes on it
- `@dependabot squash and merge` will squash and merge this PR after
your CI passes on it
- `@dependabot cancel merge` will cancel a previously requested merge
and block automerging
- `@dependabot reopen` will reopen this PR if it is closed
- `@dependabot close` will close this PR and stop Dependabot recreating
it. You can achieve the same result by closing it manually
- `@dependabot ignore this major version` will close this PR and stop
Dependabot creating any more for this major version (unless you reopen
the PR or upgrade to it yourself)
- `@dependabot ignore this minor version` will close this PR and stop
Dependabot creating any more for this minor version (unless you reopen
the PR or upgrade to it yourself)
- `@dependabot ignore this dependency` will close this PR and stop
Dependabot creating any more for this dependency (unless you reopen the
PR or upgrade to it yourself)


</details>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Travis Patterson <travis.patterson@grafana.com>

periklis committed:
operator: Provide community bundle for openshift community hub (#8881)

adityacs commented:
Small nits otherwise LGTM

adityacs commented:
Can we make the code comments simple? Mainly this one https://github.com/amankrsingh2000/loki/blob/8bf0d157fc712a54eeda81354c3864258a20b303/pkg/storage/chunk/client/ibmcloud/trusted_profile_authentication_provider.go#L36-L48

I am not against this but generally not followed in Loki. Also, it looks more like a Pythonic way of adding docstrings.

chaudum committed:
Extend scheduler queue metrics with enqueue/dequeue counters (#8891)

**What this PR does / why we need it**:

Better o11y of the scheduler. This change yields new metrics with potentially high cardinality on the scheduler.

Signed-off-by: Christian Haudum <christian.haudum@gmail.com>

darxriggs committed:
Promtail: Fix examples how to build it (#8898)

The build flags have to be provided before the package paths.

Otherwise the `build` command fails with this error:
> malformed import path "--tags=promtail_journal_enabled": leading dash

See `go help build`:
> usage: go build [-o output] [build flags] [packages]

dannykopping committed:
Distributor: add detail to stream rates failure (#8900)

**What this PR does / why we need it**:
Currently we cannot see if a single ingester is the source of the
`unable to get stream rates` failures. This change adds the client
address to the log entry.

Signed-off-by: Danny Kopping <danny.kopping@grafana.com>

dannykopping committed:
Ruler: Implement consistent rule evaluation jitter (#8896)

**What this PR does / why we need it**:
This PR replaces the previous random jitter with a consistent jitter.
While both are random, having the random jitter be applied
_consistently_ is essential for evaluating rules on a predictable
cadence.

If a rule is supposed to evaluate every minute, whether it evaluates at
(e.g.) `01:00` or `01:03.234` is irrelevant because the evaluation
_instant_ is not adjusted, so it will produce the same result whether
run at `01:00` or `01:03.234`. However, if 1000 rules are set to
evaluate at `01:00`, this will create a resource contention issue.

periklis committed:
operator: Prepare community release v0.2.0 (#8651)

abaguas committed:
azure: respect retry config before cancelling the context (#8732)

**What this PR does / why we need it**:

On GET blob operations a context timeout is set to `RequestTimeout`,
however this value is not the full timeout. This is only the first
timeout and afterwards there are retries. As a result, the retry
configuration is never used because the context is immediately
cancelled.

To fix we should set the context timeout to a value larger than the
Azure timeout after retries. Since the backoff is exponential and
randomized we need to take the value of `MaxRetryDelay` to set an upper
bound.

salvacorts committed:
Salvacorts/max querier size messaging (#8916)

**What this PR does / why we need it**:
In https://github.com/grafana/loki/pull/8670 we introduced a new limit
`max_querier_bytes_read`. When the limit was surpassed the following
error message is printed:

```
query too large to execute on a single querier, either because parallelization is not enabled, the query is unshardable, or a shard query is too big to execute: (query: %s, limit: %s). Consider adding more specific stream selectors or reduce the time range of the query
```

As pointed out in [this comment][1], a user would have a hard time
figuring out whether the cause was `parallelization is not enabled`,
`the query is unshardable` or `a shard query is too big to execute`.

This PR improves the error messaging for the `max_querier_bytes_read`
limit to raise a different error for each of the causes above.

**Which issue(s) this PR fixes**:
Followup for https://github.com/grafana/loki/pull/8670

**Special notes for your reviewer**:

**Checklist**
- [x] Reviewed the
[`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md)
guide (**required**)
- [x] Documentation added
- [x] Tests updated
- [ ] `CHANGELOG.md` updated
- [ ] Changes that require user attention or interaction to upgrade are
documented in `docs/sources/upgrading/_index.md`

[1]: https://github.com/grafana/loki/pull/8670#discussion_r1146008266

---------

Co-authored-by: Danny Kopping <danny.kopping@grafana.com>

athira-varghese committed:
feat(storage): add support for IBM cloud object storage as storage client (#8826)

**What this PR does / why we need it**:
   Add support for IBM cloud object storage as storage client
   
**Which issue(s) this PR fixes**:
Fixes NA



**Checklist**
- [x] Reviewed the
[`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md)
guide (**required**)
- [x] Documentation added
- [x] Tests updated
- [ ] `CHANGELOG.md` updated

---------

Signed-off-by: Shahul <shahulsonhal@gmail.com>
Co-authored-by: Aman Kumar Singh <amankrsingh2110@gmail.com>
Co-authored-by: Suruthi-G-K <shruthi.suruthi@gmail.com>
Co-authored-by: tareqmamari <tariq.mamari@de.ibm.com>
Co-authored-by: shahulsonhal <shahulsonhal@gmail.com>
Co-authored-by: Aditya C S <aditya.gnu@gmail.com>
Co-authored-by: Tareq Al-Maamari <tariq.mamari@gmail.com>

tareqmamari reviewed (approved):
LGTM, thanks.
